pkg/sysfs/system.go

```diff
@@ -54,37 +54,97 @@ const (
 )
 
 // System devices
-type System struct {
-	logger.Logger                 // our logger instance
-	flags         DiscoveryFlag   // system discovery flags
-	path          string          // sysfs mount point
-	packages      map[ID]*Package // physical packages
-	nodes         map[ID]*Node    // NUMA nodes
-	cpus          map[ID]*CPU     // CPUs
-	cache         map[ID]*Cache   // Cache
-	offline       IDSet           // offlined CPUs
-	isolated      IDSet           // isolated CPUs
-	threads       int             // hyperthreads per core
-}
-
-// Package is a physical package (a collection of CPUs).
-type Package struct {
+type System interface {
+	Discover(flags DiscoveryFlag) error
+	SetCpusOnline(online bool, cpus IDSet) (IDSet, error)
+	SetCPUFrequencyLimits(min, max uint64, cpus IDSet) error
+	PackageIDs() []ID
+	NodeIDs() []ID
+	CPUIDs() []ID
+	PackageCount() int
+	SocketCount() int
+	CPUCount() int
+	NUMANodeCount() int
+	ThreadCount() int
+	CPUSet() cpuset.CPUSet
+	Package(id ID) CPUPackage
+	Node(id ID) Node
+	CPU(id ID) CPU
+	Offlined() cpuset.CPUSet
+	Isolated() cpuset.CPUSet
+	discoverCache(path string) error
```

Comment:
Do these unexported functions really need to be part of the interface ? I think there is no reason to force any internal mocked test implementation to structurally mimick the discovery code so closely. And for externally mocked tests they do not matter as they are not visible.

Reply:
Right, removed them.